python/tank/platform/engine.py

```diff
@@ -1832,14 +1836,69 @@ def start_engine(engine_name, tk, context):
     Raises TankEngineInitError if an engine could not be started
     for the passed context.
     """
+    return _start_engine(engine_name, tk, None, context)
+
+
+def _restart_engine(new_context):
+
+    engine = current_engine()
+    try:
+        # Track some of the current state before restarting the engine.
+        old_context = engine.context
+
+        # Restart the engine. If we were given a new context to use,
+        # use it, otherwise restart using the same context as before.
+        new_context = new_context or engine.context
```

Comment:
Yes, I agree. It's feels like a problem with how the hook is named - maybe we can brainstorm it? 

So this hook basically runs both at engine init and context change? It's whenever a context is being associated with the engine, not necessarily when the context is changing.

It feels like we should document the different states as part of the hook docstring too. What is the meaning of:

- current_context and next_context are different
- current_context is None
- current_context and next_context are same


Maybe context setup, context init or something like that is better?

```
class EngineContextInitHook:
    def pre_context_init(self, current_context, next_context):
    def post_context_init(self, previous_context, current_context):

class EngineContextSetupHook:
    def pre_context_init(self, current_context, next_context):
    def post_context_init(self, previous_context, current_context):

class EngineContextBindHook:
    def pre_context_bind(self, current_context, next_context):
    def post_context_bind(self, previous_context, current_context):

```

Those are suggestions - not sure I really ❤️  any of them TBH... Let's discuss!